stetho/src/main/java/com/facebook/stetho/inspector/protocol/module/DOM.java

```diff
@@ -172,64 +212,245 @@ public void run() {
+          removedEvent.nodeId = mObjectIdMapper.getIdForObject(element);
+          mPeerManager.sendNotificationToPeers("DOM.childNodeRemoved", removedEvent);
+          releaseChildNodeRemovedEvent(removedEvent);
+        }
+
+        // All garbage elements should be unhooked.
+        mObjectIdMapper.removeObject(element);
+      }
+    });
+
+    // Transmit DOM changes over to Chrome
+    domUpdate.getChangedElements(new Accumulator<Object>() {
+      private final HashSet<Object> domInsertedElements = new HashSet<>();
+
+      private final NodeFactory nodeFactory = new NodeFactory() {
+        @Override
+        public Node createNodeForElement(DOMView domView, Object element) {
+          if (domUpdate.isElementChanged(element)) {
+            domInsertedElements.add(element);
+          }
+          return super.createNodeForElement(domView, element);
+        }
+      };
+
+      @Override
+      public void store(Object element) {
+        // If this returns false then it means the element was garbage, and has already been removed
+        if (!mObjectIdMapper.containsObject(element)) {
+          return;
+        }
+
+        if (domInsertedElements.contains(element)) {
+          // This element was already transmitted in its entirety by a DOM.childNodeInserted event.
+          // Trying to send any further updates about it is both unnecessary and incorrect (we'd
+          // end up with duplicated elements and really bad performance).
+          return;
+        }
+
+        final ElementInfo newElementInfo = domUpdate.getElementInfo(element);
+        final ElementInfo oldElementInfo = mShadowDOM.getElementInfo(element);
+
+        final List<Object> oldChildren = (oldElementInfo != null)
+            ? oldElementInfo.children
+            : Collections.emptyList();
+
+        final List<Object> newChildren = newElementInfo.children;
+
+        // This list is representative of Chrome's view of the DOM.
+        // We need to sync up Chrome with newChildren.
+        ChildEventingList domChildren = acquireChildEventingList(element, domUpdate);
+        for (int i = 0, N = oldChildren.size(); i < N; ++i) {
+          final Object childElement = oldChildren.get(i);
+          if (mObjectIdMapper.containsObject(childElement)) {
+            domChildren.add(childElement);
+          }
+        }
+        updateDOMChildren(domChildren, newChildren, nodeFactory);
+        releaseChildEventingList(domChildren);
+      }
+    });
+
+    domUpdate.commit();
+
+    VerifyShadowDOMMatchesDOMProvider();
+  }
+
+  // TODO: organization of test code
+  private void VerifyShadowDOMMatchesDOMProvider() {
+    final List<Object> shadowDOMList = new ArrayList<>();
+    Queue<Object> shadowDOMQueue = new ArrayDeque<>();
+    shadowDOMQueue.add(mShadowDOM.getRootElement());
+    DOMViewUtil.TraverseBFS(
+        mShadowDOM,
+        shadowDOMQueue,
+        new Function<ElementInfo, TreeTraverseAction>() {
```

Comment:
Well, at this stage it's only being used here. But it was used a lot more while I was developing this. Most instances just got pruned away at the distillation phase (e.g. ShadowDOM.UpdateImpl.getGarbageElements() couldn't use it because it has *slightly* different needs). If I don't end up using it outside of test code then I'll toss it, but it has proven to be a solid generalization.